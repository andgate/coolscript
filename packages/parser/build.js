const process = require('process')
const esbuild = require('esbuild')
const NearleyPlugin = require('esbuild-plugin-nearley')

let watch = process.argv.some((arg) => arg == '--watch')
if (watch) {
  watch = {
    onRebuild(error, result) {
      if (error) console.error('watch build failed:', error)
      else console.log('watch build succeeded:', result)
    }
  }
}

const shared = {
  entryPoints: ['src/index.ts'],
  outdir: 'lib/',
  target: 'es6',
  sourcemap: true,
  bundle: true,
  minify: false,
  tsconfig: 'tsconfig.build.json',
  plugins: [NearleyPlugin.default],
  watch
}

esbuild
  .build({
    ...shared,
    format: 'esm',
    splitting: true
  })
  .catch(() => process.exit(1))

esbuild
  .build({
    ...shared,
    format: 'cjs',
    outExtension: { '.js': '.cjs' }
  })
  .catch(() => process.exit(1))
