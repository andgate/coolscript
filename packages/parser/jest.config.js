const base = require('../../jest.config.base.js')
const { name: packageName } = require('./package.json')

module.exports = {
  ...base,
  name: packageName,
  displayName: packageName,
  moduleNameMapper: {
    '^@coolscript/parser$': '<rootDir>/src'
  },
  transform: {
    ...base.transform,
    '^.+\\.ne$': 'jest-transform-nearley'
  }
}
