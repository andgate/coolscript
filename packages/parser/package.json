{
  "name": "@coolscript/parser",
  "version": "0.0.3",
  "description": "Coolscript programming language parser.",
  "main": "./lib/index.js",
  "module": "./lib/index.cjs",
  "types": "./lib/index.d.ts",
  "publishConfig": {
    "main": "./lib/index.js",
    "module": "./lib/index.cjs",
    "typings": "lib/index.d.ts",
    "access": "public"
  },
  "files": [
    "lib/",
    "LICENSE"
  ],
  "scripts": {
    "start": "pnpm watch",
    "build": "pnpm build:lib",
    "build:lib": "pnpm clean:lib && pnpm build:grammar && pnpm build:ts && pnpm build:js",
    "build:js": "node build.js",
    "build:ts": "tsc",
    "build:grammar": "pnpm clean:grammar && nearleyc src/grammar.ne -o src/grammar.js",
    "watch": "concurrently \"pnpm:watch:*\"",
    "watch:js": "node build.js --watch",
    "watch:ts": "tsc --watch",
    "check": "tsc --noEmit",
    "prettier": "prettier --check src/ tests/",
    "prettier:fix": "prettier --write src/ tests/",
    "lint": "eslint src/** tests/** --ext .ts,.tsx",
    "lint:fix": "pnpm lint --fix",
    "verify": "pnpm prettier && pnpm lint",
    "verify:fix": "pnpm prettier:fix && pnpm lint:fix",
    "test": "jest",
    "clean": "pnpm clean:grammar && pnpm clean:lib && pnpm clean:etc",
    "clean:lib": "rimraf lib tsconfig.tsbuildinfo",
    "clean:grammar": "rimraf src/grammar.js",
    "clean:etc": "rimraf coverage .pnpm-debug.log tsconfig.tsbuildinfo",
    "clean:node_modules": "rimraf node_modules"
  },
  "author": "Gabriel Anderson",
  "license": "MIT",
  "devDependencies": {
    "@types/moo": "^0.5.5",
    "@types/nearley": "^2.11.2",
    "moo": "^0.5.1",
    "nearley": "^2.20.1"
  },
  "dependencies": {
    "@coolscript/syntax": "^0.0.3"
  },
  "gitHead": "9b91bf0df0bb73d74440b9a1d191fd5c91c6af93"
}
