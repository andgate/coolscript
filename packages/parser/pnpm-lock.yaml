lockfileVersion: 5.3

specifiers:
  '@types/moo': ^0.5.5
  '@types/nearley': ^2.11.2
  esbuild-plugin-nearley: ^1.0.3
  jest-transform-nearley: ^2.0.0
  moo: ^0.5.1
  nearley: ^2.20.1

devDependencies:
  '@types/moo': 0.5.5
  '@types/nearley': 2.11.2
  esbuild-plugin-nearley: 1.0.3_nearley@2.20.1
  jest-transform-nearley: 2.0.0_nearley@2.20.1
  moo: 0.5.1
  nearley: 2.20.1

packages:

  /@types/moo/0.5.5:
    resolution: {integrity: sha512-eXQpwnkI4Ntw5uJg6i2PINdRFWLr55dqjuYQaLHNjvqTzF14QdNWbCbml9sza0byyXNA0hZlHtcdN+VNDcgVHA==}
    dev: true

  /@types/nearley/2.11.2:
    resolution: {integrity: sha512-jeyIDNBxxyWyEk6HemDC+t32b4fxthVsgWDxf88qD2WpX0QpOyY8JvA80AElPTBGRUsO0EKnr6OeVOjK3ZDdnA==}
    dev: true

  /commander/2.20.3:
    resolution: {integrity: sha512-GpVkmM8vF2vQUkj2LvZmD35JxeJOLCwJ9cUkugyk2nuhbv3+mJvpLYYt+0+USMxE+oj+ey/lJEnhZw75x/OMcQ==}
    dev: true

  /discontinuous-range/1.0.0:
    resolution: {integrity: sha1-44Mx8IRLukm5qctxx3FYWqsbxlo=}
    dev: true

  /esbuild-plugin-nearley/1.0.3_nearley@2.20.1:
    resolution: {integrity: sha512-put4IL3TkSRETj1jvqPGXRfpXmtZiuD5Er41UgKTlZsH7cjoHbPtXP/2FDBdt0D9BB5pWZRnOQET+0NQXOqoBA==}
    engines: {node: '>=16', npm: '>=8', pnpm: '>=6'}
    peerDependencies:
      esbuild: ^0.14
      nearley: ^2
    dependencies:
      nearley: 2.20.1
    dev: true

  /jest-transform-nearley/2.0.0_nearley@2.20.1:
    resolution: {integrity: sha512-tjLgaASIKmB0zVFdSyQoJZQDmcyIpExkWF8O+0ph9nWE+3a3YsWTDTC0Uuc20Erz8QJ8RcL2U0c+USr696K4HQ==}
    peerDependencies:
      nearley: ^2
    dependencies:
      nearley: 2.20.1
    dev: true

  /moo/0.5.1:
    resolution: {integrity: sha512-I1mnb5xn4fO80BH9BLcF0yLypy2UKl+Cb01Fu0hJRkJjlCRtxZMWkTdAtDd5ZqCOxtCkhmRwyI57vWT+1iZ67w==}
    dev: true

  /nearley/2.20.1:
    resolution: {integrity: sha512-+Mc8UaAebFzgV+KpI5n7DasuuQCHA89dmwm7JXw3TV43ukfNQ9DnBH3Mdb2g/I4Fdxc26pwimBWvjIw0UAILSQ==}
    hasBin: true
    dependencies:
      commander: 2.20.3
      moo: 0.5.1
      railroad-diagrams: 1.0.0
      randexp: 0.4.6
    dev: true

  /railroad-diagrams/1.0.0:
    resolution: {integrity: sha1-635iZ1SN3t+4mcG5Dlc3RVnN234=}
    dev: true

  /randexp/0.4.6:
    resolution: {integrity: sha512-80WNmd9DA0tmZrw9qQa62GPPWfuXJknrmVmLcxvq4uZBdYqb1wYoKTmnlGUchvVWe0XiLupYkBoXVOxz3C8DYQ==}
    engines: {node: '>=0.12'}
    dependencies:
      discontinuous-range: 1.0.0
      ret: 0.1.15
    dev: true

  /ret/0.1.15:
    resolution: {integrity: sha512-TTlYpa+OL+vMMNG24xSlQGEJ3B/RzEfUlLct7b5G/ytav+wPrplCpVMFuwzXbkecJrb6IYo1iFb0S9v37754mg==}
    engines: {node: '>=0.12'}
    dev: true
