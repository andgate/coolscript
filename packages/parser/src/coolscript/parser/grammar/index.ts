/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Grammar } from 'nearley'
// @ts-ignore
import grammarRules from './grammar.ne'

export const coolscriptGrammar: Grammar = Grammar.fromCompiled(grammarRules)
