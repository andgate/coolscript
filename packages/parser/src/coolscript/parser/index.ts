export * from './coolscript-parser'
export * from './token'
