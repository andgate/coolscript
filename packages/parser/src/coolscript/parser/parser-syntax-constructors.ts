import * as Syntax from '@coolscript/syntax'
import { Span, Merge } from '@coolscript/syntax'
import { Token } from './token'

export function VariableDeclaration(
  lhs: Token,
  rhs: Syntax.Term
): Syntax.VariableDeclaration {
  return Syntax.VariableDeclaration(lhs.text, rhs, Merge(lhs.span, rhs.span))
}

export function NullValue(t: Token): Syntax.NullValue {
  return Syntax.NullValue(t.span)
}

export function NumberValue(t: Token): Syntax.NumberValue {
  const num = parseFloat(t.text)
  return Syntax.NumberValue(num, t.span)
}

export function StringValue(t: Token): Syntax.StringValue {
  const sourceText = t.text
  const text = sourceText.substring(1, sourceText.length - 1)
  return Syntax.StringValue(text, t.span)
}

export function TrueValue(t: Token): Syntax.BooleanValue {
  return Syntax.BooleanValue(true, t.span)
}
export function FalseValue(t: Token): Syntax.BooleanValue {
  return Syntax.BooleanValue(false, t.span)
}

export function ErrorTerm(msg: string, span: Span): Syntax.ErrorTerm {
  return Syntax.ErrorTerm(msg, span)
}

export function ValueTerm(value: Syntax.AtomicValue): Syntax.ValueTerm {
  return Syntax.ValueTerm(value, value.span)
}

export function VariableTerm(variableToken: Token): Syntax.VariableTerm {
  return Syntax.VariableTerm(variableToken.text, variableToken.span)
}

export function AssignmentTerm(
  lhs: Token,
  rhs: Syntax.Term
): Syntax.AssignmentTerm {
  return Syntax.AssignmentTerm(lhs.text, rhs, Merge(lhs.span, rhs.span))
}

export function LambdaTerm(
  args: Array<Syntax.Variable>,
  body: Syntax.Term,
  span: Span
): Syntax.LambdaTerm {
  return Syntax.LambdaTerm(args, body, span)
}

export function CallTerm(
  func: Syntax.Term,
  args: Array<Syntax.Term>,
  span: Span
): Syntax.CallTerm {
  return Syntax.CallTerm(func, args, span)
}

export function LetTerm(
  declarations: Array<Syntax.Declaration>,
  body: Syntax.Term,
  span: Span
): Syntax.LetTerm {
  return Syntax.LetTerm(declarations, body, span)
}

export function ParentheticalTerm(
  term: Syntax.Term,
  span: Span
): Syntax.ParentheticalTerm {
  return Syntax.ParentheticalTerm(term, span)
}

export function ArrayTerm(
  elements: Array<Syntax.Term>,
  span: Span
): Syntax.ArrayTerm {
  return Syntax.ArrayTerm(elements, span)
}

export function ObjectTerm(
  entries: Syntax.ObjectMap<Syntax.Term>,
  span: Span
): Syntax.ObjectTerm {
  return Syntax.ObjectTerm(entries, span)
}

export function MemberAccessTerm(
  object: Syntax.Term,
  memberToken: Token
): Syntax.MemberAccessTerm {
  return Syntax.MemberAccessTerm(
    object,
    memberToken.text,
    Merge(object.span, memberToken.span)
  )
}

export function IndexAccessTerm(
  array: Syntax.Term,
  index: Syntax.Term,
  span: Span
): Syntax.IndexAccessTerm {
  return Syntax.IndexAccessTerm(array, index, span)
}

export function DoTerm(
  block: Syntax.BlockStatement,
  span: Span
): Syntax.DoTerm {
  return Syntax.DoTerm(block, span)
}

export function ConditionalTerm(
  condition: Syntax.Term,
  body: Syntax.Term,
  branch: Syntax.BranchTerm,
  span: Span
): Syntax.ConditionalTerm {
  return Syntax.ConditionalTerm(condition, body, branch, span)
}

export function ElifTerm(
  condition: Syntax.Term,
  body: Syntax.Term,
  branch: Syntax.BranchTerm,
  span: Span
): Syntax.ElifTerm {
  return Syntax.ElifTerm(condition, body, branch, span)
}

export function ElseTerm(body: Syntax.Term, span: Span): Syntax.ElseTerm {
  return Syntax.ElseTerm(body, span)
}

export function AssignmentStatement(
  lhs: Token,
  rhs: Syntax.Term
): Syntax.AssignmentStatement {
  return Syntax.AssignmentStatement(lhs.text, rhs, Merge(lhs.span, rhs.span))
}

export function CallStatement(
  func: Syntax.Term,
  args: Array<Syntax.Term>,
  span: Span
): Syntax.CallStatement {
  return Syntax.CallStatement(func, args, span)
}

export function ReturnStatement(
  result: Syntax.Term,
  span: Span
): Syntax.ReturnStatement {
  return Syntax.ReturnStatement(result, span)
}

export function BlockStatement(
  statements: Array<Syntax.Statement>,
  span: Span
): Syntax.BlockStatement {
  return Syntax.BlockStatement(statements, span)
}

export function IfStatement(
  condition: Syntax.Term,
  body: Syntax.Statement,
  branch: Syntax.BranchStatement | null,
  span: Span
): Syntax.IfStatement {
  return Syntax.IfStatement(condition, body, branch, span)
}

export function ElifStatement(
  condition: Syntax.Term,
  body: Syntax.Statement,
  branch: Syntax.BranchStatement | null,
  span: Span
): Syntax.ElifStatement {
  return Syntax.ElifStatement(condition, body, branch, span)
}

export function ElseStatement(
  body: Syntax.Statement,
  span: Span
): Syntax.ElseStatement {
  return Syntax.ElseStatement(body, span)
}

export function WhileStatement(
  condition: Syntax.Term,
  body: Syntax.Statement,
  span: Span
): Syntax.WhileStatement {
  return Syntax.WhileStatement(condition, body, span)
}

export function DoWhileStatement(
  body: Syntax.Statement,
  condition: Syntax.Term,
  span: Span
): Syntax.DoWhileStatement {
  return Syntax.DoWhileStatement(body, condition, span)
}

export function ForStatement(
  declarations: Array<Syntax.VariableDeclaration> | null | undefined,
  condition: Syntax.Term,
  update: Syntax.Term,
  body: Syntax.Statement,
  span: Span
): Syntax.ForStatement {
  return Syntax.ForStatement(declarations, condition, update, body, span)
}
