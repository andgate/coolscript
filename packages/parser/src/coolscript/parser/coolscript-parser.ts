import * as nearley from 'nearley'
import { Term } from '@coolscript/syntax'
import { coolscriptGrammar } from './grammar'

export type ParseResult = {
  term: Term | null
  warnings: string[]
  errors?: Error[]
}

function ParseSuccess(term: Term, warnings: string[]): ParseResult {
  return { term, warnings }
}

function ParseFail(warnings: string[], ...errors: Error[]): ParseResult {
  return { term: null, errors, warnings }
}

export function parse(src: string): ParseResult {
  const parser = new nearley.Parser(coolscriptGrammar)
  const warnings = []
  try {
    parser.feed(src)
  } catch (error) {
    let errorMsg = ''
    if (error && error.stack && error.message) {
      errorMsg = ` Error message: ${error.message}`
    }
    const msg = `Parser failed!${errorMsg}`
    const parserFailedError = new Error(msg)
    return ParseFail(warnings, parserFailedError)
  }

  if (
    !parser.results ||
    !Array.isArray(parser.results) ||
    parser.results.length == 0
  ) {
    const noParseResultsError = new Error('No parse results were recovered.')
    return ParseFail(warnings, noParseResultsError)
  }

  if (parser.results.length > 1) {
    const ambiguousGrammarWarning = new Error(
      'Ambiguous grammar encountered! Resolving first result.'
    )
    warnings.push(ambiguousGrammarWarning)
  }
  const term: Term = parser.results[0] as Term // must assume at this point
  return ParseSuccess(term, warnings)
}
