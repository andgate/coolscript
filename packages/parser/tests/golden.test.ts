import { examples } from '@coolscript/examples'
import { parse } from '@coolscript/parser'

describe('Parser Golden Test Suite', () => {
  test.each(examples)('Example %#', (example) => {
    const parseResult = parse(example)
    expect(parseResult).toMatchSnapshot()
  })
})
