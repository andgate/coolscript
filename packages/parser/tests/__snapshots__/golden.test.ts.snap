// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example 1 1`] = `"{\\"term\\":{\\"tag\\":\\"LetTerm\\",\\"declarations\\":[{\\"tag\\":\\"VariableDeclaration\\",\\"variable\\":\\"id\\",\\"body\\":{\\"tag\\":\\"LambdaTerm\\",\\"args\\":[\\"x\\"],\\"body\\":{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"x\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":3,\\"end\\":3},\\"column\\":{\\"start\\":15,\\"end\\":15}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":3,\\"end\\":3},\\"column\\":{\\"start\\":8,\\"end\\":15}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":3,\\"end\\":3},\\"column\\":{\\"start\\":3,\\"end\\":15}}}},{\\"tag\\":\\"VariableDeclaration\\",\\"variable\\":\\"constant\\",\\"body\\":{\\"tag\\":\\"LambdaTerm\\",\\"args\\":[\\"a\\",\\"b\\"],\\"body\\":{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"a\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":4,\\"end\\":4},\\"column\\":{\\"start\\":24,\\"end\\":24}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":4,\\"end\\":4},\\"column\\":{\\"start\\":14,\\"end\\":24}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":4,\\"end\\":4},\\"column\\":{\\"start\\":3,\\"end\\":24}}}},{\\"tag\\":\\"VariableDeclaration\\",\\"variable\\":\\"createFoo\\",\\"body\\":{\\"tag\\":\\"LambdaTerm\\",\\"args\\":[\\"x\\"],\\"body\\":{\\"tag\\":\\"ObjectTerm\\",\\"entries\\":{\\"name\\":{\\"tag\\":\\"ArrayTerm\\",\\"elements\\":[{\\"tag\\":\\"ValueTerm\\",\\"value\\":{\\"tag\\":\\"StringValue\\",\\"str\\":\\"F\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":12,\\"end\\":14}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":12,\\"end\\":14}}}},{\\"tag\\":\\"ValueTerm\\",\\"value\\":{\\"tag\\":\\"StringValue\\",\\"str\\":\\"o\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":17,\\"end\\":19}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":17,\\"end\\":19}}}},{\\"tag\\":\\"ValueTerm\\",\\"value\\":{\\"tag\\":\\"StringValue\\",\\"str\\":\\"o\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":22,\\"end\\":24}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":22,\\"end\\":24}}}},{\\"tag\\":\\"ValueTerm\\",\\"value\\":{\\"tag\\":\\"StringValue\\",\\"str\\":\\"b\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":27,\\"end\\":29}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":27,\\"end\\":29}}}},{\\"tag\\":\\"ValueTerm\\",\\"value\\":{\\"tag\\":\\"StringValue\\",\\"str\\":\\"a\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":32,\\"end\\":34}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":32,\\"end\\":34}}}},{\\"tag\\":\\"ValueTerm\\",\\"value\\":{\\"tag\\":\\"StringValue\\",\\"str\\":\\"r\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":37,\\"end\\":39}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":37,\\"end\\":39}}}}],\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":6,\\"end\\":6},\\"column\\":{\\"start\\":11,\\"end\\":40}}}},\\"value\\":{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"x\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":7,\\"end\\":7},\\"column\\":{\\"start\\":12,\\"end\\":12}}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":5,\\"end\\":8},\\"column\\":{\\"start\\":22,\\"end\\":3}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":5,\\"end\\":8},\\"column\\":{\\"start\\":15,\\"end\\":3}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":5,\\"end\\":8},\\"column\\":{\\"start\\":3,\\"end\\":3}}}},{\\"tag\\":\\"VariableDeclaration\\",\\"variable\\":\\"bar\\",\\"body\\":{\\"tag\\":\\"ValueTerm\\",\\"value\\":{\\"tag\\":\\"NumberValue\\",\\"num\\":999,\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":9,\\"end\\":9},\\"column\\":{\\"start\\":9,\\"end\\":11}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":9,\\"end\\":9},\\"column\\":{\\"start\\":9,\\"end\\":11}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":9,\\"end\\":9},\\"column\\":{\\"start\\":3,\\"end\\":11}}}}],\\"body\\":{\\"tag\\":\\"DoTerm\\",\\"block\\":{\\"tag\\":\\"BlockStatement\\",\\"statements\\":[{\\"tag\\":\\"AssignmentStatement\\",\\"lhs\\":\\"baz\\",\\"rhs\\":{\\"tag\\":\\"CallTerm\\",\\"func\\":{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"id\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":11,\\"end\\":11},\\"column\\":{\\"start\\":9,\\"end\\":10}}}},\\"args\\":[{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"bar\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":11,\\"end\\":11},\\"column\\":{\\"start\\":12,\\"end\\":14}}}}],\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":11,\\"end\\":11},\\"column\\":{\\"start\\":9,\\"end\\":15}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":11,\\"end\\":11},\\"column\\":{\\"start\\":3,\\"end\\":15}}}},{\\"tag\\":\\"CallStatement\\",\\"func\\":{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"constant\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":12,\\"end\\":12},\\"column\\":{\\"start\\":3,\\"end\\":10}}}},\\"args\\":[{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"baz\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":12,\\"end\\":12},\\"column\\":{\\"start\\":12,\\"end\\":14}}}},{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"bar\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":12,\\"end\\":12},\\"column\\":{\\"start\\":17,\\"end\\":19}}}}],\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":12,\\"end\\":12},\\"column\\":{\\"start\\":3,\\"end\\":20}}}},{\\"tag\\":\\"AssignmentStatement\\",\\"lhs\\":\\"clientName\\",\\"rhs\\":{\\"tag\\":\\"MemberAccessTerm\\",\\"object\\":{\\"tag\\":\\"CallTerm\\",\\"func\\":{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"createFoo\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":13,\\"end\\":13},\\"column\\":{\\"start\\":16,\\"end\\":24}}}},\\"args\\":[{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"baz\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":13,\\"end\\":13},\\"column\\":{\\"start\\":26,\\"end\\":28}}}}],\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":13,\\"end\\":13},\\"column\\":{\\"start\\":16,\\"end\\":29}}}},\\"member\\":\\"name\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":13,\\"end\\":13},\\"column\\":{\\"start\\":16,\\"end\\":34}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":13,\\"end\\":13},\\"column\\":{\\"start\\":3,\\"end\\":34}}}},{\\"tag\\":\\"ReturnStatement\\",\\"result\\":{\\"tag\\":\\"VariableTerm\\",\\"variable\\":\\"clientName\\",\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":14,\\"end\\":14},\\"column\\":{\\"start\\":10,\\"end\\":19}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":14,\\"end\\":14},\\"column\\":{\\"start\\":3,\\"end\\":19}}}}],\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":10,\\"end\\":15},\\"column\\":{\\"start\\":7,\\"end\\":1}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":10,\\"end\\":15},\\"column\\":{\\"start\\":4,\\"end\\":1}}}},\\"ann\\":{\\"span\\":{\\"type\\":\\"Span\\",\\"line\\":{\\"start\\":2,\\"end\\":15},\\"column\\":{\\"start\\":1,\\"end\\":1}}}},\\"warnings\\":[{}]}"`;
